addons/mass_mailing/controllers/main.py

```diff
@@ -21,21 +21,53 @@ def mailing(self, mailing_id, email=None, res_id=None, token="", **post):
-
             right_token = mailing._unsubscribe_token(res_id, email)
             if not consteq(str(token), right_token):
                 raise exceptions.AccessDenied()
-            mailing.update_opt_out(email, res_ids, True)
-            return _('You have been unsubscribed successfully')
+
+            if mailing.mailing_model_real == 'mail.mass_mailing.contact':
+                # Unsubscribe directly + Let the user choose his subscriptions
+                mailing.update_opt_out(email, mailing.contact_list_ids.ids, True)
+
+                contacts = request.env['mail.mass_mailing.contact'].sudo().search([('email', '=', email)])
+                list_ids = contacts.mapped('opt_out_list_ids')
+                # In many user are found : if user is opt_out on the list with contact_id 1 but not with contact_id 2,
+                # assume that the user is not opt_out on both
+                opt_out_list_ids = list_ids.filtered(lambda rel: rel.opt_out).mapped('list_id')
+                opt_in_list_ids = list_ids.filtered(lambda rel: not rel.opt_out).mapped('list_id')
+                opt_out_list_ids = set([list.id for list in opt_out_list_ids if list not in opt_in_list_ids])
+
+                unique_list_ids = set([list.list_id.id for list in list_ids])
+                list_ids = request.env['mail.mass_mailing.list'].sudo().browse(unique_list_ids)
+                unsubscribed_list = ', '.join(str(list.name) for list in mailing.contact_list_ids if list.is_public)
+                return request.render('mass_mailing.page_unsubscribe', {
+                    'contacts': contacts,
+                    'list_ids': list_ids,
+                    'opt_out_list_ids': opt_out_list_ids,
+                    'unsubscribed_list': unsubscribed_list,
+                    'email': email,
+                    'mailing_id': mailing_id,
+                    'res_id': res_id,
+                    'show_blacklist_button': request.env['ir.config_parameter'].sudo().get_param('mass_mailing.show_blacklist_buttons'),
+                    })
+            else:
+                recipient_name = 'partner' if mailing.mailing_model_real == 'res.partner' else 'lead'
+                request.env['mail.blacklist'].sudo()._add(email, """The %s asked to not be contacted anymore 
+                    using an unsubscribe link.""" % recipient_name)
+                return request.render('mass_mailing.page_unsubscribed', {
+                    'email': email,
+                    'mailing_id': mailing_id,
+                    'res_id': res_id,
+                    'show_blacklist_button': request.env['ir.config_parameter'].sudo().get_param(
+                        'mass_mailing.show_blacklist_buttons'),
+                    })
+
+    @http.route('/mail/mailing/unsubscribe', type='json', auth='none')
+    def unsubscribe(self, mailing_id, opt_in_ids, opt_out_ids, email):
+        mailing = request.env['mail.mass_mailing'].sudo().browse(mailing_id)
```

Comment:
Isn't it a security hole ? A route available for everyone that perform a sudo then update records based on email without any check ...